[lldb] Check if C++ interop and embedded Swift are enabled on CU instead of whole module

augusto2112 opened the pull request:
Checking if C++ interop is enabled can be extremely slow if the project is big enough. We have changed this to checking only the compile unit instead, but one usage of the old API slipped by.

rdar://147009063

augusto2112 committed:
[lldb] Check if interop is enabled only on compile unit

When deciding whether to format a C++ type as a Swiftified type or not,
checking if Swift/C++ interop is enabled on the entire target can be
very expensive for big projects, and may not swiftify types that should
be.

rdar://117708944
(cherry picked from commit fe06ebc96d94076cdf6bad599556ba3dbbd97f8d)

augusto2112 committed:
[lldb] Check if C++ interop is enabled on CU instead of whole module

Checking if C++ interop is enabled can be extremely slow if the project
is big enough. We have changed this to checking only the compile unit
instead, but one usage of the old API slipped by.

rdar://147009063

augusto2112 commented:
@swift-ci test

adrian-prantl reviewed (approved):
That makes a lot more sense than the copy&paste error that was there before.

augusto2112 commented:
@swift-ci test

augusto2112 committed:
[lldb] Check if Embedded Swift is enabled only on the compile unit

Reading debug info of an entire module to check whether embedded Swift
is enabled or not can be very slow. This patch replaces that check with
only checking the compile unit instead.

rdar://147009063

augusto2112 commented:
@swift-ci test

kastiglione reviewed (commented):
The code looks reasonable to me, but I don't have the knowledge to know whether this prevents the expensive flag checks.

augusto2112 commented:
@swift-ci test macOS

augusto2112 commented:
@swift-ci test macos

adrian-prantl commented on lldb/source/Plugins/Language/Swift/SwiftLanguage.cpp:
```diff
@@ -1237,6 +1234,29 @@ SwiftLanguage::GetHardcodedSynthetics() {
+          lldb::SymbolContextItem::eSymbolContextModule);
+
+      // If there is a compile unit, use that to check if C++ interop should be
+      // enabled. If there is no compiler unit, use the module. If neither
```
compiler -> compile

adrian-prantl commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -2653,7 +2650,7 @@ static bool ShouldEnableCXXInterop(CompileUnit *cu) {
         if (cu->GetLanguage() == eLanguageTypeSwift)
           for (const char *arg : args.GetArgumentArrayRef())
-            if (strcmp(arg, "-enable-experimental-cxx-interop") == 0)
+            if (strcmp(arg, flag) == 0)
```
Do we not have a DWARF attribute for this?

augusto2112 replied:
We don't

adrian-prantl commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -2663,6 +2660,15 @@ static bool ShouldEnableCXXInterop(CompileUnit *cu) {
   return false;
 }
 
+/// Determine whether this CU was compiled with C++ interop enabled.
```
this comment should be in the header
